Add RustCrypto benchmarks

AxelNennker opened the pull request:
Add RustCrypto benchmarks. These are benchmarks for the sha2 crate.
Also rust_crypto is unmaintained and not benchmarked anymore.

I agree to license my contributions to each file under the terms given at the top of each file I changed.

Signed-off-by: Axel Nennker <axel.nennker@telekom.de>

AxelNennker committed:
Add RustCrypto benchmarks and uncomment rust_crypto because it is unmaintained

I agree to license my contributions to each file under the terms given at the top of each file I changed.

Signed-off-by: Axel Nennker <axel.nennker@telekom.de>

briansmith committed:
Remove Octavo.

Octavo seems to be unmaintained now.

briansmith committed:
Replace `cargo_all` and `cargo_all.bat` with Cargo workspace.

Cargo workspaces make these scripts obsolete.

briansmith committed:
Clarify linking story.

briansmith committed:
Update dependencies to latest releases.

briansmith committed:
Fix minor typo in README.md.

briansmith commented:
Please drop the removal of the original rust-crypto benchmarks. There are still a lot of crates depending on the original rust-crypto and it may be helpful to current users to compare the performance of what they're currently using to what they might switch to.

Please rebase on the current master, which switched to using a cargo workspace to replace the `cargo_all` script.

briansmith commented on RustCrypto/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+[dependencies]
+#sha2 = { git = "https://github.com/RustCrypto/hashes" }
+sha2 = { git = "https://github.com/RustCrypto/hashes", features = ["asm"] }
+
```
Everything below this point should be removed since this was all moved to the top-level Cargo.toml.

AxelNennker replied:
done

briansmith commented on RustCrypto/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+
+[dependencies]
+#sha2 = { git = "https://github.com/RustCrypto/hashes" }
+sha2 = { git = "https://github.com/RustCrypto/hashes", features = ["asm"] }
```
Let's just use the latest release on crates.io, like we now do for the other crates.

Probably it makes sense for the benchmarks to have an `asm` feature that, when provided, enables the `asm` feature of this crate, and otherwise doesn't enable it. This should be documented in the README.

AxelNennker replied:
Done

briansmith commented on RustCrypto/RustCrypto.rs:
```diff
@@ -0,0 +1,163 @@
+    rust_crypto_digest_benches!(sha1, crypto_bench::SHA1_BLOCK_LEN,
+                                crypto_bench::SHA1_OUTPUT_LEN,
+                                sha1::Sha1::new());
+*/
```
Why is this commented out? I think we should just remove it.

AxelNennker replied:
I kept it for future contributors (or myself) as a reminder what to do next.
Removed the commented out code for now.

briansmith commented on RustCrypto/RustCrypto.rs:
```diff
@@ -0,0 +1,163 @@
+                                sha2::Sha512::default());
+}
+
+/*
```
Please remove all the commented-out code.

AxelNennker replied:
done

briansmith reviewed (changes requested):
After you update this PR, please ask somebody familiar with RustCrypto post their own review of this PR. In particular, besides a general overview, they should verify whether you are using RustCrypto correctly. They should also pay particular attention to whether the benchmarks is measuring the exact same thing as the analogous benchmarks for the other crates.

briansmith committed:
Fix Travis CI.

AxelNennker committed:
add RustCrypto to workbench

I agree to license my contributions to each file under the terms given at the top of each file I changed.

AxelNennker committed:
Remove stuff that was moved to toplevel Cargo.toml

I agree to license my contributions to each file under the terms given at the top of each file I changed.

AxelNennker committed:
use sha2 version from crates.io not the latest from github

I agree to license my contributions to each file under the terms given at the top of each file I changed.

AxelNennker committed:
Removed comments

I agree to license my contributions to each file under the terms given at the top of each file I changed.

AxelNennker commented:
I fail to resolve the merge conflict. Would you @briansmith please do this and then merge this PR?
When I try to do this strange things happen with my clone. I am about to throw it away and start from scratch.

briansmith commented:
Thank you. Unfortunately, I just don't have time to work on this project so I'm going to stop accepting contributions for a while.
